File: core/js/src/span_identifier_v2.ts
// Mirror of core/py/src/braintrust_core/span_identifier_v2.py.

import * as uuid from "uuid";
import { ParentExperimentIds, ParentProjectLogIds } from "./object";
import { SpanComponentsV1 } from "./span_identifier_v1";
import { z } from "zod";

function tryMakeUuid(s: string): { bytes: Buffer; isUUID: boolean } {
  try {
    const ret = uuid.parse(s);
    if (ret.length !== 16) {
      throw new Error();
    }
    return { bytes: Buffer.from(ret), isUUID: true };
  } catch (e) {
    return { bytes: Buffer.from(s, "utf-8"), isUUID: false };
  }
}

const ENCODING_VERSION_NUMBER = 2;

const INVALID_ENCODING_ERRMSG = `SpanComponents string is not properly encoded. This library only supports encoding versions up to ${ENCODING_VERSION_NUMBER}. Please make sure the SDK library used to decode the SpanComponents is at least as new as any library used to encode it.`;
// If you change this, make sure to change the method used to read/write integer
// bytes to a buffer, from writeInt32BE.
const INTEGER_ENCODING_NUM_BYTES = 4;

export enum SpanObjectTypeV2 {
  EXPERIMENT = 1,
  PROJECT_LOGS = 2,
}

const SpanObjectTypeV2EnumSchema = z.nativeEnum(SpanObjectTypeV2);

export class SpanRowIdsV2 {
  public rowId: string;
  public spanId: string;
  public rootSpanId: string;

  constructor(args: { rowId: string; spanId: string; rootSpanId: string }) {
    this.rowId = args.rowId;
    this.spanId = args.spanId;
    this.rootSpanId = args.rootSpanId;

    if (!this.rowId) {
      throw new Error("rowId must be nonempty string");
    }
    if (!this.spanId) {
      throw new Error("spanId must be nonempty string");
    }
    if (!this.rootSpanId) {
      throw new Error("rootSpanId must be nonempty string");
    }
  }

  public toObject() {
    return {
      rowId: this.rowId,
      spanId: this.spanId,
      rootSpanId: this.rootSpanId,
    };
  }
}

export class SpanComponentsV2 {
  public objectType: SpanObjectTypeV2;
  public objectId: string | undefined;
  public computeObjectMetadataArgs: Record<string, any> | undefined;
  public rowIds: SpanRowIdsV2 | undefined;

  constructor(args: {
    objectType: SpanObjectTypeV2;
    objectId?: string;
    computeObjectMetadataArgs?: Record<string, any>;
    rowIds?: SpanRowIdsV2;
  }) {
    this.objectType = args.objectType;
    this.objectId = args.objectId;
    this.computeObjectMetadataArgs = args.computeObjectMetadataArgs;
    this.rowIds = args.rowIds;

    if (!(this.objectId || this.computeObjectMetadataArgs)) {
      throw new Error(
        "Must provide either objectId or computeObjectMetadataArgs",
      );
    }
  }

  public toStr(): string {
    const allBuffers: Array<Buffer> = [];

    const { bytes: rowIdBytes, isUUID: rowIdIsUUID } = this.rowIds
      ? tryMakeUuid(this.rowIds.rowId)
      : { bytes: Buffer.from(""), isUUID: false };

    allBuffers.push(
      Buffer.from([
        ENCODING_VERSION_NUMBER,
        this.objectType,
        this.objectId ? 1 : 0,
        this.computeObjectMetadataArgs ? 1 : 0,
        this.rowIds ? 1 : 0,
        rowIdIsUUID ? 1 : 0,
      ]),
    );

    if (this.objectId) {
      const { bytes: objectIdBytes, isUUID: objectIdIsUUID } = tryMakeUuid(
        this.objectId,
      );
      if (!objectIdIsUUID) {
        throw new Error("object_id component must be a valid UUID");
      }
      allBuffers.push(objectIdBytes);
    }

    if (this.computeObjectMetadataArgs) {
      const computeObjectMetadataBytes = Buffer.from(
        JSON.stringify(this.computeObjectMetadataArgs),
        "utf-8",
      );
      const serializedLenBytes = Buffer.alloc(INTEGER_ENCODING_NUM_BYTES);
      serializedLenBytes.writeInt32BE(computeObjectMetadataBytes.length);
      allBuffers.push(serializedLenBytes, computeObjectMetadataBytes);
    }

    if (this.rowIds) {
      const { bytes: spanIdBytes, isUUID: spanIdIsUUID } = tryMakeUuid(
        this.rowIds.spanId,
      );
      if (!spanIdIsUUID) {
        throw new Error("span_id component must be a valid UUID");
      }
      const { bytes: rootSpanIdBytes, isUUID: rootSpanIdIsUUID } = tryMakeUuid(
        this.rowIds.rootSpanId,
      );
      if (!rootSpanIdIsUUID) {
        throw new Error("root_span_id component must be a valid UUID");
      }
      allBuffers.push(spanIdBytes, rootSpanIdBytes, rowIdBytes);
    }

    return Buffer.concat(allBuffers).toString("base64");
  }

  public static fromStr(s: string): SpanComponentsV2 {
    try {
      const rawBytes = Buffer.from(s, "base64");

      if (rawBytes[0] < ENCODING_VERSION_NUMBER) {
        const spanComponentsOld = SpanComponentsV1.fromStr(s);
        return new SpanComponentsV2({
          objectType: SpanObjectTypeV2EnumSchema.parse(
            spanComponentsOld.objectType,
          ),
          objectId: spanComponentsOld.objectId,
          rowIds: spanComponentsOld.rowIds
            ? new SpanRowIdsV2({
                rowId: spanComponentsOld.rowIds.rowId,
                spanId: spanComponentsOld.rowIds.spanId,
                rootSpanId: spanComponentsOld.rowIds.rootSpanId,
              })
            : undefined,
        });
      }

      if (rawBytes[0] !== ENCODING_VERSION_NUMBER) {
        throw new Error();
      }
      const objectType = SpanObjectTypeV2EnumSchema.parse(rawBytes[1]);
      for (let i = 2; i < 6; ++i) {
        if (![0, 1].includes(rawBytes[i])) {
          throw new Error();
        }
      }
      const hasObjectId = rawBytes[2] == 1;
      const hasComputeObjectMetadataArgs = rawBytes[3] == 1;
      const hasRowId = rawBytes[4] == 1;
      const rowIdIsUUID = rawBytes[5] == 1;

      let byteCursor = 6;
      let objectId: string | undefined = undefined;
      if (hasObjectId) {
        const nextByteCursor = byteCursor + 16;
        objectId = uuid.stringify(
          rawBytes.subarray(byteCursor, nextByteCursor),
        );
        byteCursor = nextByteCursor;
      }

      let computeObjectMetadataArgs: Record<string, any> | undefined;
      if (hasComputeObjectMetadataArgs) {
        let nextByteCursor = byteCursor + INTEGER_ENCODING_NUM_BYTES;
        const serializedLenBytes = rawBytes.readInt32BE(byteCursor);
        byteCursor = nextByteCursor;
        nextByteCursor = byteCursor + serializedLenBytes;
        computeObjectMetadataArgs = JSON.parse(
          rawBytes.subarray(byteCursor, nextByteCursor).toString("utf-8"),
        );
        byteCursor = nextByteCursor;
      }

      const rowIds = (() => {
        if (!hasRowId) {
          return undefined;
        }
        let nextByteCursor = byteCursor + 16;
        const spanId = uuid.stringify(
          rawBytes.subarray(byteCursor, nextByteCursor),
        );
        byteCursor = nextByteCursor;
        nextByteCursor = byteCursor + 16;
        const rootSpanId = uuid.stringify(
          rawBytes.subarray(byteCursor, nextByteCursor),
        );
        byteCursor = nextByteCursor;
        const rowId = rowIdIsUUID
          ? uuid.stringify(rawBytes.subarray(byteCursor))
          : rawBytes.subarray(byteCursor).toString("utf-8");
        return new SpanRowIdsV2({ rowId, spanId, rootSpanId });
      })();

      return new SpanComponentsV2({
        objectType,
        objectId,
        computeObjectMetadataArgs,
        rowIds,
      });
    } catch (e) {
      throw new Error(INVALID_ENCODING_ERRMSG);
    }
  }

  public objectIdFields(): ParentExperimentIds | ParentProjectLogIds {
    if (!this.objectId) {
      throw new Error(
        "Impossible: cannot invoke `object_id_fields` unless SpanComponentsV2 is initialized with an `object_id`",
      );
    }
    switch (this.objectType) {
      case SpanObjectTypeV2.EXPERIMENT:
        return { experiment_id: this.objectId };
      case SpanObjectTypeV2.PROJECT_LOGS:
        return { project_id: this.objectId, log_id: "g" };
      default:
        throw new Error("Impossible");
    }
  }

  public toObject() {
    return {
      objectType: this.objectType,
      objectId: this.objectId,
      computeObjectMetadataArgs: this.computeObjectMetadataArgs,
      rowIds: this.rowIds?.toObject(),
    };
  }
}

File: py/src/braintrust/span_identifier_v2.py
# Serialization format for capturing all relevant information about a span
# necessary for distributed logging / tracing. Meant to be passed around as an
# opaque string.

import base64
import dataclasses
import json
from enum import Enum
from uuid import UUID

from .span_identifier_v1 import SpanComponentsV1


def _try_make_uuid(s):
    try:
        ret = UUID(s).bytes
        assert len(ret) == 16
        return ret, True
    except Exception:
        return s.encode("utf-8"), False


ENCODING_VERSION_NUMBER = 2
INTEGER_ENCODING_NUM_BYTES = 4
INTEGER_ENCODING_BYTEORDER = "big"

INVALID_ENCODING_ERRMSG = f"SpanComponents string is not properly encoded. This library only supports encoding versions up to {ENCODING_VERSION_NUMBER}. Please make sure the SDK library used to decode the SpanComponents is at least as new as any library used to encode it."


class SpanObjectTypeV2(Enum):
    EXPERIMENT = 1
    PROJECT_LOGS = 2

    def __str__(self):
        return {SpanObjectTypeV2.EXPERIMENT: "experiment", SpanObjectTypeV2.PROJECT_LOGS: "project_logs"}[self]


@dataclasses.dataclass
class SpanRowIdsV2:
    row_id: str
    span_id: str
    root_span_id: str

    def __post_init__(self):
        assert isinstance(self.row_id, str)
        assert isinstance(self.span_id, str)
        assert isinstance(self.root_span_id, str)
        assert self.row_id
        assert self.span_id
        assert self.root_span_id


@dataclasses.dataclass
class SpanComponentsV2:
    object_type: SpanObjectTypeV2
    object_id: str | None = None
    compute_object_metadata_args: dict | None = None
    row_ids: SpanRowIdsV2 | None = None

    def __post_init__(self):
        assert isinstance(self.object_type, SpanObjectTypeV2)
        assert self.object_id or self.compute_object_metadata_args, (
            "Must provide either object_id or compute_object_metadata_args"
        )
        if self.object_id:
            assert isinstance(self.object_id, str)
        else:
            assert isinstance(self.compute_object_metadata_args, dict)
        if self.row_ids is not None:
            assert isinstance(self.row_ids, SpanRowIdsV2)

    def to_str(self) -> str:
        # Our binary object format is as follows:
        #   - Byte 0 encodes the version number of the encoded string. This is
        #   used to check for incompatibilities with previous iterations.
        #   - Byte 1 encodes the SpanObjectTypeV2.
        #   - Byte 2 (0 or 1) encodes whether or not we have an object_id.
        #   - Byte 3 (0 or 1) encodes whether or not we have
        #   compute_object_metadata_args.
        #   - Byte 4 (0 or 1) describes whether or not the (row_id,
        #   span_id, root_span_id) triple is present.
        #   - Byte 5 (0 or 1) describes whether or not the row_id component
        #   is a UUID. If not, it is assumed to be a utf-8 encoded string.
        #   - If [byte 2] == 1, the next 16 bytes encode the object_id as a UUID.
        #   - If [byte 3] == 1, the next [INTEGER_ENCODING_NUM_BYTES] bytes
        #   encode the length of the serialized compute_object_metadata_args. The next
        #   [length] bytes contain the serialized compute_object_metadata_args.
        #   - If [byte 4] == 1, the next 32 bytes encode the span_id +
        #   root_span_id as UUIDs.
        #   - If the row triple is present, the remaining bytes encode the
        #   row_id either as UUID or as UTF-8.

        if self.row_ids:
            row_id_bytes, row_id_is_uuid = _try_make_uuid(self.row_ids.row_id)
        else:
            row_id_bytes, row_id_is_uuid = None, False

        raw_bytes = bytes(
            [
                ENCODING_VERSION_NUMBER,
                self.object_type.value,
                1 if self.object_id else 0,
                1 if self.compute_object_metadata_args else 0,
                1 if self.row_ids else 0,
                1 if row_id_is_uuid else 0,
            ]
        )

        if self.object_id:
            object_id_bytes, object_id_is_uuid = _try_make_uuid(self.object_id)
            if not object_id_is_uuid:
                raise Exception("object_id component must be a valid UUID")
            raw_bytes += object_id_bytes

        if self.compute_object_metadata_args:
            compute_object_metadata_bytes = bytes(json.dumps(self.compute_object_metadata_args).encode())
            serialized_len_bytes = len(compute_object_metadata_bytes).to_bytes(
                INTEGER_ENCODING_NUM_BYTES, byteorder=INTEGER_ENCODING_BYTEORDER
            )
            raw_bytes += serialized_len_bytes
            raw_bytes += compute_object_metadata_bytes

        if self.row_ids:
            span_id_bytes, span_id_is_uuid = _try_make_uuid(self.row_ids.span_id)
            if not span_id_is_uuid:
                raise Exception("span_id component must be a valid UUID")
            root_span_id_bytes, root_span_id_is_uuid = _try_make_uuid(self.row_ids.root_span_id)
            if not root_span_id_is_uuid:
                raise Exception("root_span_id component must be a valid UUID")
            raw_bytes += span_id_bytes
            raw_bytes += root_span_id_bytes
            raw_bytes += row_id_bytes

        return base64.b64encode(raw_bytes).decode()

    @staticmethod
    def from_str(s: str) -> "SpanComponentsV2":
        try:
            raw_bytes = base64.b64decode(s.encode())

            if raw_bytes[0] < ENCODING_VERSION_NUMBER:
                span_components_old = SpanComponentsV1.from_str(s)
                object_type = SpanObjectTypeV2(span_components_old.object_type.value)
                if span_components_old.row_ids:
                    row_ids = SpanRowIdsV2(
                        row_id=span_components_old.row_ids.row_id,
                        span_id=span_components_old.row_ids.span_id,
                        root_span_id=span_components_old.row_ids.root_span_id,
                    )
                else:
                    row_ids = None
                return SpanComponentsV2(
                    object_type=object_type, object_id=span_components_old.object_id, row_ids=row_ids
                )

            assert raw_bytes[0] == ENCODING_VERSION_NUMBER
            object_type = SpanObjectTypeV2(raw_bytes[1])
            for i in range(2, 6):
                assert raw_bytes[i] in [0, 1]
            has_object_id = raw_bytes[2]
            has_compute_object_metadata_args = raw_bytes[3]
            has_row_id = raw_bytes[4] == 1
            row_id_is_uuid = raw_bytes[5] == 1

            byte_cursor = 6
            if has_object_id:
                next_byte_cursor = byte_cursor + 16
                object_id = str(UUID(bytes=raw_bytes[byte_cursor:next_byte_cursor]))
                byte_cursor = next_byte_cursor
            else:
                object_id = None

            if has_compute_object_metadata_args:
                next_byte_cursor = byte_cursor + INTEGER_ENCODING_NUM_BYTES
                serialized_len_bytes = int.from_bytes(
                    raw_bytes[byte_cursor:next_byte_cursor], byteorder=INTEGER_ENCODING_BYTEORDER
                )
                byte_cursor = next_byte_cursor
                next_byte_cursor = byte_cursor + serialized_len_bytes
                compute_object_metadata_args = json.loads(raw_bytes[byte_cursor:next_byte_cursor].decode())
                byte_cursor = next_byte_cursor
            else:
                compute_object_metadata_args = None

            if has_row_id:
                next_byte_cursor = byte_cursor + 16
                span_id = str(UUID(bytes=raw_bytes[byte_cursor:next_byte_cursor]))
                byte_cursor = next_byte_cursor
                next_byte_cursor = byte_cursor + 16
                root_span_id = str(UUID(bytes=raw_bytes[byte_cursor:next_byte_cursor]))
                byte_cursor = next_byte_cursor
                if row_id_is_uuid:
                    row_id = str(UUID(bytes=raw_bytes[byte_cursor:]))
                else:
                    row_id = raw_bytes[byte_cursor:].decode("utf-8")
                row_ids = SpanRowIdsV2(row_id=row_id, span_id=span_id, root_span_id=root_span_id)
            else:
                row_ids = None

            return SpanComponentsV2(
                object_type=object_type,
                object_id=object_id,
                compute_object_metadata_args=compute_object_metadata_args,
                row_ids=row_ids,
            )
        except Exception:
            raise Exception(INVALID_ENCODING_ERRMSG)

    def object_id_fields(self):
        if not self.object_id:
            raise Exception(
                "Impossible: cannot invoke `object_id_fields` unless SpanComponentsV2 is initialized with an `object_id`"
            )
        if self.object_type == SpanObjectTypeV2.EXPERIMENT:
            return dict(experiment_id=self.object_id)
        elif self.object_type == SpanObjectTypeV2.PROJECT_LOGS:
            return dict(project_id=self.object_id, log_id="g")
